### PR TITLE
MEP_oM: Add objects for initial oM definition

### LINKED ISSUE #577: MEP_oM: Create Initial Prototype
Following workshop with @al-fisher, proposing categorizing MEP objects into four primary categories/namespaces:

``` C#
BH.oM.MEP.Equipment
BH.oM.MEP.Parts
BH.oM.MEP.Systems
BH.oM.MEP.Fixtures
```
Sketch of Equipment/Part breakdown: 

image

### PR DESCRIPTION
**Pull request by FraserGreenroyd**
### Issues addressed by this PR
Fixes #577 

### Test files
N/A

### Changelog
 - Add Equipment objects to MEP oM
 - Add Part objects to MEP oM

### Additional comments
Initial prototype ready for further discussion being led by @kayleighhoude

### CODE REVIEW

File: MEP_oM/Parts/HeatingCoil.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    public class HeatingCoil : BHoMObject, IPart
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public double SensibleCapacity { get; set; } = 0.0;
+        public double EnteringDryBulbAirTemperature { get; set; } = 0.0;
+        public double LeavingDryBulbAirTemperature { get; set; } = 0.0;
+        public double EnteringWaterTemperature { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Parts/HeatingCoil.cs]: Fluid, as might now always be water, could be some mix with anti-freeze

File: MEP_oM/Equipment/AirHandlingUnit.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        public string Type { get; set; } = "";
+        public double TotalSupplyAirFlow { get; set; } = 0.0;
+        public double TotalSupplyAirExternalStaticPressure { get; set; } = 0.0;
+        public double TotalReturnAirFlow { get; set; } = 0.0;
+        public double TotalReturnAirExternalStaticPressure { get; set; } = 0.0;
+
+        public double TotalDesignOutdoorAirFlow { get; set; } = 0.0;
+        public double DemandControlledVentilationMinimumOutdoorAirFlow { get; set; } = 0.0;
+        public double TotalOutdoorAirFlowExternalStaticPressure { get; set; } = 0.0;
+        public double TotalReliefAirFlow { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Equipment/AirHandlingUnit.cs]: `TotalReliefAirFlow` to be `TotalExhaustAirFlow`

so  we have:
OutdoorAir - external air that is entering AHU
SupplyAir - air we supply to space after going thought AHU
ExhaustAir - air that is leaving space into AHU
ExtractAir - air that is leaving AHU
*ReturnAir - if we  use mixing box this is air from space that is mixing with OutdoorAir

File: MEP_oM/Equipment/AirHandlingUnit.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public string Type { get; set; } = "";
+        public double TotalSupplyAirFlow { get; set; } = 0.0;
+        public double TotalSupplyAirExternalStaticPressure { get; set; } = 0.0;
+        public double TotalReturnAirFlow { get; set; } = 0.0;
+        public double TotalReturnAirExternalStaticPressure { get; set; } = 0.0;
+
+        public double TotalDesignOutdoorAirFlow { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Equipment/AirHandlingUnit.cs]: *here we call Design - is not better call for all this Flow `..SpeecifiedFlow` so this is clear what we use and I guess we could add Design as additional to show what was design on

File: MEP_oM/Equipment/AirHandlingUnit.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        public double DemandControlledVentilationMinimumOutdoorAirFlow { get; set; } = 0.0;
+        public double TotalOutdoorAirFlowExternalStaticPressure { get; set; } = 0.0;
+        public double TotalReliefAirFlow { get; set; } = 0.0;
+        public double TotalReliefExternalStaticPressure { get; set; } = 0.0;
+
+        public bool SupplyAirEconomiser { get; set; } = false;
+        public bool WaterEconomiser { get; set; } = false;
+
```
Comment (michaldengusiak) [inline MEP_oM/Equipment/AirHandlingUnit.cs]: add
ReferenceDensity [kg/m3]
ReferenceAltitudeAbovSeaLevel [m]

File: MEP_oM/Equipment/AirHandlingUnit.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        public string Type { get; set; } = "";
+        public double TotalSupplyAirFlow { get; set; } = 0.0;
+        public double TotalSupplyAirExternalStaticPressure { get; set; } = 0.0;
+        public double TotalReturnAirFlow { get; set; } = 0.0;
+        public double TotalReturnAirExternalStaticPressure { get; set; } = 0.0;
+
+        public double TotalDesignOutdoorAirFlow { get; set; } = 0.0;
+        public double DemandControlledVentilationMinimumOutdoorAirFlow { get; set; } = 0.0;
+        public double TotalOutdoorAirFlowExternalStaticPressure { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Equipment/AirHandlingUnit.cs]: not sure if this is needed,

File: MEP_oM/Equipment/AirHandlingUnit.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+{
+    public class AirHandlingUnit : BHoMObject, IEquipment
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public string Type { get; set; } = "";
+        public double TotalSupplyAirFlow { get; set; } = 0.0;
+        public double TotalSupplyAirExternalStaticPressure { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Equipment/AirHandlingUnit.cs]: I would call it SupplyExternalStaticPressure

File: MEP_oM/Equipment/AirHandlingUnit.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        public double TotalReturnAirFlow { get; set; } = 0.0;
+        public double TotalReturnAirExternalStaticPressure { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Equipment/AirHandlingUnit.cs]: not sure if this is needed but I would call it ReturnExternalStaticPressure,

File: MEP_oM/Equipment/AirHandlingUnit.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+        public double TotalDesignOutdoorAirFlow { get; set; } = 0.0;
+        public double DemandControlledVentilationMinimumOutdoorAirFlow { get; set; } = 0.0;
+        public double TotalOutdoorAirFlowExternalStaticPressure { get; set; } = 0.0;
+        public double TotalReliefAirFlow { get; set; } = 0.0;
+        public double TotalReliefExternalStaticPressure { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Equipment/AirHandlingUnit.cs]: I would call it ExhaustExternalStaticPressure,

File: MEP_oM/Parts/CoolingCoil.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        public double SensibleCapacity { get; set; } = 0.0;
+        public double EnteringDryBulbAirTemperature { get; set; } = 0.0;
+        public double EnteringWetBulbAirTemperature { get; set; } = 0.0;
+        public double LeavingDryBulbAirTemperature { get; set; } = 0.0;
+        public double LeavingWetBulbAirTemperature { get; set; } = 0.0;
+        public double EnteringWaterTemperature { get; set; } = 0.0;
+        public double LeavingWaterTemperature { get; set; } = 0.0;
+        public double PressureDrop { get; set; } = 0.0;
+        public int NumberOfRows { get; set; } = 0;
+
```
Comment (michaldengusiak) [inline MEP_oM/Parts/CoolingCoil.cs]: Add:
CoolingMediumPressureDrop [Pa]
FaceVelocity [m/s]

File: MEP_oM/Parts/ElectricalConnector.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace BH.oM.MEP.Parts
+{
+    public class ElectricalConnector : BHoMObject, IPart
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public double MotorHorsePower { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Parts/ElectricalConnector.cs]: Power [W]  * we use SI

File: MEP_oM/Parts/Fan.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public double FlowRate { get; set; } = 0.0;
+        public double ExternalStaticPressure { get; set; } = 0.0;
+        public double Speed { get; set; } = 0.0;
+        public string DriveType { get; set; } = "";
+        public string SpeedControl { get; set; } = "";
+        public double BrakeHorsePower { get; set; } = 0.0;
+        public double HorsePower { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Parts/Fan.cs]: Power [W] SI units

File: MEP_oM/Parts/Fan.cs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        public double Efficiency { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Parts/Fan.cs]: TotalEfficiency [%]

File: MEP_oM/Parts/Fan.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
```
Comment (michaldengusiak) [inline MEP_oM/Parts/Fan.cs]: Add
MotorEfficiency [%]
TemperatureRise [degC]

File: MEP_oM/Parts/Filter.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+namespace BH.oM.MEP.Parts
+{
+    public class Filter : BHoMObject, IPart
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public string Type { get; set; } = "";
+        public int MERVRating { get; set; } = 0;
```
Comment (michaldengusiak) [inline MEP_oM/Parts/Filter.cs]: FilterClass (F7 etc)

File: MEP_oM/Parts/Filter.cs [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        public double InitialPressureDrop { get; set; } = 0.0;
+        public double ReplacementPressureDrop { get; set; } = 0.0;
+        public double Area { get; set; } = 0.0;
```
Comment (michaldengusiak) [inline MEP_oM/Parts/Filter.cs]: FaceArea [m2]

File: MEP_oM/Parts/Filter.cs [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
```
Comment (michaldengusiak) [inline MEP_oM/Parts/Filter.cs]: Add:
 FaceVelocity [m/s]
DesignPressureDrop [Pa]

**Comment by michaldengusiak** [comment]
thanks for implementing my comments :)

**Comment by rwemay** [comment]
Great chat today thanks @agnieszkaszymczyk and @kayleighhoude , as discussed, if @agnieszkaszymczyk  is happy, I suggest we merge this so we can move forward. @michaldengusiak great comments! I think as most items are additive, the suggestion is to raise these in a separate issue and address in the next wave, to keep the minimum possible for this commit. Will wait for your review before approve/merge @agnieszkaszymczyk

File: MEP_oM/Fragments/Identity.cs [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+namespace BH.oM.MEP.Fragments
+{
+    public class IdentityFragment : IBHoMFragment
+    {
+        public string Manufacturer { get; set; } = "";
+        public string ModelNumber { get; set; } = "";
+        public string Location { get; set; } = "";
+        public string Service { get; set; } = "";
+        public List<string> Remarks { get; set; } = new List<string>();
+    }
```
Comment (agnieszkaszymczyk) [inline MEP_oM/Fragments/Identity.cs]: Could we please add name/type just for ease of identification ? Unless its already added else where .

File: MEP_oM/Parts/CoolingCoil.cs [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+    public class CoolingCoil : BHoMObject, IPart
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public double TotalCapacity { get; set; } = 0.0;
+        public double SensibleCapacity { get; set; } = 0.0;
+        public double EnteringDryBulbAirTemperature { get; set; } = 0.0;
+        public double EnteringWetBulbAirTemperature { get; set; } = 0.0;
```
[the remark this answers is not in the document]
Comment (agnieszkaszymczyk) [reply MEP_oM/Parts/CoolingCoil.cs]: As discussed - dry / wet bulb is fine as long as RH or enthalpy can be calculated from the built in data base . Would be good to have an option what the user wants to see i.e. weather it is dry bulb and wet bulb or dry bulb and RH.

File: MEP_oM/Fragments/Identity.cs (hunk 17 above)
Comment (FraserGreenroyd) [reply MEP_oM/Fragments/Identity.cs]: I can confirm 'name' is a property inherited from 'BHoMObject' so will exist on all these objects should they be needed :smile:

File: MEP_oM/Parts/CoolingCoil.cs [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        public double LeavingDryBulbAirTemperature { get; set; } = 0.0;
+        public double LeavingWetBulbAirTemperature { get; set; } = 0.0;
+        public double EnteringWaterTemperature { get; set; } = 0.0;
+        public double LeavingWaterTemperature { get; set; } = 0.0;
+        public double PressureDrop { get; set; } = 0.0;
+        public int NumberOfRows { get; set; } = 0;
```
Comment (agnieszkaszymczyk) [inline MEP_oM/Parts/CoolingCoil.cs]: Agree - FaceVelocity  needs to be added . All the rest can be added as required.

File: MEP_oM/Parts/EnergyWheel.cs [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public double OutdoorSummerEnteringDryBulbTemperature { get; set; } = 0.0;
+        public double OutdoorSummerEnteringWetBulbTemperature { get; set; } = 0.0;
+        public double OutdoorSummerLeavingDryBulbTemperature { get; set; } = 0.0;
+        public double OutdoorSummerLeavingWetBulbTemperature { get; set; } = 0.0;
+
+        public double ReliefSummerEnteringDryBulbTemperature { get; set; } = 0.0;
+        public double ReliefSummerEnteringWetBulbTemperature { get; set; } = 0.0;
```
Comment (agnieszkaszymczyk) [inline MEP_oM/Parts/EnergyWheel.cs]: As discussed - naming of each  particular parameters will always be slightly different across regions and will depend on people preferences . I think the term exhaust is commonly known?  common language as much as possible please and if needed and requested by majority we can modify as required.

File: MEP_oM/Parts/EnergyWheel.cs [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+        public double ReliefSummerLeavingDryBulbTemperature { get; set; } = 0.0;
+        public double ReliefSummerLeavingWetBulbTemperature { get; set; } = 0.0;
+
+        public double SummerSensibleEffectiveness { get; set; } = 0.0;
+        public double SummerTotalEffectiveness { get; set; } = 0.0;
+
+        public double OutdoorWinterEnteringDryBulbTemperature { get; set; } = 0.0;
+        public double OutdoorWinterEnteringWetBulbTemperature { get; set; } = 0.0;
```
[the remark this answers is not in the document]
Comment (agnieszkaszymczyk) [reply MEP_oM/Parts/EnergyWheel.cs]: Can we replace Effectivness with Efficiency ?

File: MEP_oM/Parts/EnergyWheel.cs [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+using BH.oM.Base;
+
+namespace BH.oM.MEP.Parts
+{
+    public class EnergyWheel : BHoMObject, IPart
+    {
+        /***************************************************/
```
Comment (agnieszkaszymczyk) [inline MEP_oM/Parts/EnergyWheel.cs]: Do we need to add the type of the energy recovery component ? I belive there will be more then energywheel ?

File: MEP_oM/Parts/Fan.cs (hunk 12 above)
Comment (agnieszkaszymczyk) [reply MEP_oM/Parts/Fan.cs]: Agree needs to be added.

File: MEP_oM/Parts/Fan.cs (hunk 13 above)
Comment (agnieszkaszymczyk) [reply MEP_oM/Parts/Fan.cs]: I would leave those out as you can find them in the fan selection data sheet. We dont really need to put all the info that is in the AHU spec  - and if really needed can be always added.

File: MEP_oM/Parts/Fan.cs (hunk 11 above)
Comment (agnieszkaszymczyk) [inline MEP_oM/Parts/Fan.cs]: Do we need to add any od the accoustic info - might be usefull for the attenuators selection .

File: MEP_oM/Parts/HeatingCoil.cs (hunk 1 above)
Comment (agnieszkaszymczyk) [inline MEP_oM/Parts/HeatingCoil.cs]: Do we need to identyfy 2 different coils - FrostCoil and MainHeatingCoil

File: MEP_oM/Parts/Fan.cs (hunk 11 above)
Comment (agnieszkaszymczyk) [reply MEP_oM/Parts/Fan.cs]: Do we show SFP anywhere ?

File: MEP_oM/Fragments/Geometry.cs [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+
+using BH.oM.Geometry;
+
+namespace BH.oM.MEP.Fragments
+{
+    public class GeometryFragment : IBHoMFragment
+    {
+        public IGeometry Geometry { get; set; } = null;
+    }
+}
```
Comment (agnieszkaszymczyk) [inline MEP_oM/Fragments/Geometry.cs]: Would be good to see the overall length / width / height /  and weight based on the LOD  plus the access sections between different components .

File: MEP_oM/Parts/IPart.cs [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+using BH.oM.Base;
+
+namespace BH.oM.MEP.Parts
+{
+    public interface IPart : IBHoMObject
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
```
Comment (agnieszkaszymczyk) [inline MEP_oM/Parts/IPart.cs]: Will the motorized damper or humidifier be included under the Parts ?

**Comment by rwemay** [comment]
@agnieszkaszymczyk  thanks for the comments. As discussed and agreed, @kayleighhoude  please raise a separate issue (or broken into separate issues if necessary) summarising the comments from @agnieszkaszymczyk  and @michaldengusiak for further development/conclusion in the next (or subsequent) PR. Merging this as the initial seed then, to allow further prototyping. Please include test scripts for subsequent PR's.

**Pull request merged**